core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java

```diff
@@ -209,20 +212,31 @@ public Object visitReplaceCommand(InvocationContext ctx, ReplaceCommand command)
 
    @Override
    public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
-      return invokeNextThenAccept(ctx, command, (rCtx, rCommand, rv) -> {
+      return invokeNextThenAccept(ctx, command, handlePutMapCommandReturn);
+   }
+
+   protected InvocationSuccessAction handlePutMapCommandReturn() {
+      return (rCtx, rCommand, rv) -> {
          PutMapCommand putMapCommand = (PutMapCommand) rCommand;
          if (!isStoreEnabled(putMapCommand) || rCtx.isInTxScope())
             return;
 
-         Map<Object, Object> map = putMapCommand.getMap();
-         for (Object key : map.keySet()) {
-            if (isProperWriter(rCtx, putMapCommand, key)) {
-               storeEntry(rCtx, key, putMapCommand);
-            }
+         List<MarshalledEntry> allWriterEntries = new ArrayList<>();
+         List<MarshalledEntry> nonSharedEntries = new ArrayList<>();
+         for (Object key : putMapCommand.getMap().keySet()) {
+            if (skipSharedStores(rCtx, key, (FlagAffectedCommand) rCommand))
+               nonSharedEntries.add(createMarshalledEntry(rCtx, key));
+            else
+               allWriterEntries.add(createMarshalledEntry(rCtx, key));
          }
+
+         long flags = ((FlagAffectedCommand)rCommand).getFlagsBitSet();
+         persistenceManager.writeBatchToAllNonTxStores(allWriterEntries, BOTH, flags);
+         persistenceManager.writeBatchToAllNonTxStores(nonSharedEntries, PRIVATE, flags);
```

Comment:
`writeBatchToAllNonTxStores` already filters the entries it writes, I think we can add another check in there instead of copying the collection (not necessarily another `filter` call, because I'm not sure the filter allocation is escaped).